### PR TITLE
Add support for running DebugConfigurations with serverReadyAction

### LINKED ISSUE #186455: Extend serverReadyAction to support startDebugging with provided DebugConfiguration
`serverReadyAction` only seems to support [`name`](https://github.com/microsoft/vscode/blob/0f8598bf4664686876f64731b08b3ecdefcf69d8/extensions/debug-server-ready/package.json#L150-L154) for starting a new browser debug session. 

This feature request is to extend it to support a new entry called `config` that will be used for the `vscode.debug.startDebugging` call.

E.g.
```jsonc
"name": {
   "type": "object",
   "markdownDescription": "%debug.server.ready.config.description%",
   "default": {
      "name": "Launch Browser",
      "type": "edge",
      "request": "launch"
   },
   "properties": {
       /* vscode.DebugConfiguration */
   }
}, 
```

### PR DESCRIPTION
**Pull request by WardenGnaw**
This PR adds 'config' as an option for `serverReadyAction`. It runs similar to 'name', but if a configuration is generated on the fly, its easier to inject the debug configurations with the serverReadyAction. This modifies the startNamedDebugSession to startDebugSession that takes in a vscode.DebugConfiguration.

Resolves: #186455

### CODE REVIEW

**Comment by roblourens** [comment]
Did copilot write that 🤔

**Comment by WardenGnaw** [comment]
The second half, yes 😅

I wanted to see what it would generate.

File: extensions/debug-server-ready/src/extension.ts [hunk 1]
```diff
@@ -246,6 +251,38 @@ class ServerReadyDetector extends vscode.Disposable {
 		});
 	}
 
+	private async startConfigDebugSession(session: vscode.DebugSession, config: vscode.DebugConfiguration) {
```
Comment (roblourens) [inline extensions/debug-server-ready/src/extension.ts:254]: Any way to share logic with `startNamedDebugSession`?

Comment (WardenGnaw) [reply extensions/debug-server-ready/src/extension.ts:254]: I had one throught:

```
	private async startConfigDebugSession(session: vscode.DebugSession, name: string, config?: vscode.DebugConfiguration) {
		const args = session.configuration.serverReadyAction as ServerReadyAction;
		if (!args.killOnServerStop) {
			await vscode.debug.startDebugging(session.workspaceFolder, config ?? name);
			return;
		}

		const cts = new vscode.CancellationTokenSource();
		const newSessionPromise = this.catchStartedDebugSession(x => x.name === name, cts.token);

		if (!await vscode.debug.startDebugging(session.workspaceFolder, config ?? name)) {
			cts.cancel();
			cts.dispose();
			return;
		}

		const createdSession = await newSessionPromise;
		cts.dispose();

		if (!createdSession) {
			return;
		}

		const stopListener = this.onDidSessionStop(async () => {
			stopListener.dispose();
			this.disposables.delete(stopListener);
			await vscode.debug.stopDebugging(createdSession);
		});
		this.disposables.add(stopListener);
	}
```

Thoughts?

Comment (roblourens) [reply extensions/debug-server-ready/src/extension.ts:254]: Fine with me

**Comment by WardenGnaw** [comment]
The new AI generated summary with the new commit

-------------------
This pull request to `extensions/debug-server-ready` adds a new feature that allows a debug configuration to be launched when the server is ready. The changes include modifications to the `startNamedDebugSession` method in `extension.ts` to use a new `startDebugSession` method that can take a debug configuration object as an argument, in addition to the debug configuration name. A new `config` property has been added to the `ServerReadyAction` interface to represent a debug configuration object that can be launched when the server is ready. These changes improve the flexibility and customization of debug sessions.

* `extensions/debug-server-ready/src/extension.ts`: Modified the `startNamedDebugSession` method to use a new `startDebugSession` method that can take a debug configuration object as an argument, in addition to the debug configuration name. A new `config` property has been added to the `ServerReadyAction` interface to represent a debug configuration object that can be launched when the server is ready. (Fb352528)
* <a href="diffhunk://#diff-862a879e83f24473ac5c7196cf1671d006f3e3ab8f60837f079fb22f5cd9e442R161-R204">`extensions/debug-server-ready/package.json`</a>: Added a new debug configuration that can be launched when the server is ready, with properties such as configuration type, request type, and session termination included in the `serverReadyAction` array.

**Comment by WardenGnaw** [comment]
@roblourens Is there anything else thats needs to be changed to get this PR merged?

**Review by roblourens — APPROVED** [review]

File: extensions/debug-server-ready/package.json [hunk 2]
```diff
@@ -158,6 +158,50 @@
...
+                      "killOnServerStop": false
+                    },
+                    "required": [
+                      "config"
+                    ],
+                    "properties": {
+                      "action": {
+                        "type": "string",
+                        "enum": [
+                          "startDebugging"
```
Comment (rzhao271) [inline extensions/debug-server-ready/package.json:181]: Does this mean the only valid option for now is startDebugging?

Comment (roblourens) [reply extensions/debug-server-ready/package.json:181]: This is an extension of `serverReadyAction` to support a different type when using `startDebugging`. There are other options (the json schema is hard to parse here but they are there). `startDebugging` is the only one that is relevant in this scenario.

**Review by rzhao271 — APPROVED** [review]
